File: app/src/main/java/com/bopr/android/smailer/PermissionsHelper.java
package com.bopr.android.smailer;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.PermissionInfo;

import androidx.appcompat.app.AlertDialog;
import androidx.core.app.ActivityCompat;

import com.bopr.android.smailer.util.SharedPreferencesWrapper;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import static android.Manifest.permission.ACCESS_COARSE_LOCATION;
import static android.Manifest.permission.ACCESS_FINE_LOCATION;
import static android.Manifest.permission.PROCESS_OUTGOING_CALLS;
import static android.Manifest.permission.READ_CONTACTS;
import static android.Manifest.permission.READ_PHONE_STATE;
import static android.Manifest.permission.READ_SMS;
import static android.Manifest.permission.RECEIVE_SMS;
import static android.Manifest.permission.SEND_SMS;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static androidx.core.app.ActivityCompat.shouldShowRequestPermissionRationale;
import static androidx.core.content.ContextCompat.checkSelfPermission;
import static com.bopr.android.smailer.Settings.PREF_EMAIL_CONTENT;
import static com.bopr.android.smailer.Settings.PREF_EMAIL_TRIGGERS;
import static com.bopr.android.smailer.Settings.PREF_MARK_SMS_AS_READ;
import static com.bopr.android.smailer.Settings.VAL_PREF_EMAIL_CONTENT_CONTACT;
import static com.bopr.android.smailer.Settings.VAL_PREF_EMAIL_CONTENT_LOCATION;
import static com.bopr.android.smailer.Settings.VAL_PREF_TRIGGER_IN_CALLS;
import static com.bopr.android.smailer.Settings.VAL_PREF_TRIGGER_IN_SMS;
import static com.bopr.android.smailer.Settings.VAL_PREF_TRIGGER_MISSED_CALLS;
import static com.bopr.android.smailer.Settings.VAL_PREF_TRIGGER_OUT_CALLS;
import static com.bopr.android.smailer.Settings.VAL_PREF_TRIGGER_OUT_SMS;
import static com.bopr.android.smailer.util.AndroidUtil.checkPermission;
import static com.bopr.android.smailer.util.TagFormatter.formatter;
import static com.google.common.collect.Iterables.toArray;

/**
 * Responsible for permissions checking.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class PermissionsHelper implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static Logger log = LoggerFactory.getLogger("PermissionsHelper");

    private static final String WRITE_SMS = "android.permission.WRITE_SMS";

    private static int nextRequestResult = 200;

    private int requestResultCode = nextRequestResult++;
    private Activity activity;
    private Settings settings;
    private Map<String, Integer> items = new HashMap<>();

    @SuppressWarnings({"deprecation", "RedundantSuppression"}) // TODO: 06.02.2020 deprecated
    protected PermissionsHelper(Activity activity, Settings settings) {
        this.activity = activity;
        this.settings = settings;
        this.settings.registerChangeListener(this);

        items.put(RECEIVE_SMS, R.string.permission_rationale_receive_sms);
        items.put(WRITE_SMS, R.string.permission_rationale_write_sms);
        items.put(READ_SMS, R.string.permission_rationale_read_sms);
        items.put(READ_PHONE_STATE, R.string.permission_rationale_phone_state);
        items.put(PROCESS_OUTGOING_CALLS, R.string.permission_rationale_outgoing_call);
        items.put(READ_CONTACTS, R.string.permission_rationale_read_contacts);
        items.put(ACCESS_COARSE_LOCATION, R.string.permission_rationale_coarse_location);
        items.put(ACCESS_FINE_LOCATION, R.string.permission_rationale_fine_location);
        items.put(SEND_SMS, R.string.permission_rationale_send_sms);
    }

    public void dispose() {
        settings.unregisterChangeListener(this);
        log.debug("Disposed");
    }

    public void checkAll() {
        log.debug("Checking all");

        doCheck(items.keySet());
    }

    /**
     * To be added into activity's onRequestPermissionsResult()
     */
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == requestResultCode) {
            Set<String> deniedPermissions = new HashSet<>();

            for (int i = 0; i < grantResults.length; i++) {
                if (grantResults[i] != PERMISSION_GRANTED) {
                    deniedPermissions.add(permissions[i]);
                }
            }

            if (!deniedPermissions.isEmpty()) {
                onPermissionsDenied(deniedPermissions);
                showDenialImpact();
            }
        }
    }

    @Override
    @SuppressWarnings({"deprecation", "RedundantSuppression"}) // TODO: 06.02.2020 deprecated
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Set<String> deniedPermissions = new HashSet<>();

        switch (key) {
            case PREF_EMAIL_TRIGGERS:
                Set<String> triggers = settings.getStringSet(PREF_EMAIL_TRIGGERS);
                if (triggers.contains(VAL_PREF_TRIGGER_IN_SMS)) {
                    deniedPermissions.add(RECEIVE_SMS);
                }
                if (triggers.contains(VAL_PREF_TRIGGER_OUT_SMS)) {
                    deniedPermissions.add(READ_SMS);
                }
                if (triggers.contains(VAL_PREF_TRIGGER_IN_CALLS) ||
                        triggers.contains(VAL_PREF_TRIGGER_MISSED_CALLS)) {
                    deniedPermissions.add(READ_PHONE_STATE);
                }
                if (triggers.contains(VAL_PREF_TRIGGER_OUT_CALLS)) {
                    deniedPermissions.add(PROCESS_OUTGOING_CALLS);
                }
                break;
            case PREF_EMAIL_CONTENT:
                Set<String> content = settings.getStringSet(PREF_EMAIL_CONTENT);
                if (content.contains(VAL_PREF_EMAIL_CONTENT_CONTACT)) {
                    deniedPermissions.add(READ_CONTACTS);
                }
                if (content.contains(VAL_PREF_EMAIL_CONTENT_LOCATION)) {
                    deniedPermissions.add(ACCESS_COARSE_LOCATION);
                    deniedPermissions.add(ACCESS_FINE_LOCATION);
                }
                break;
            case PREF_MARK_SMS_AS_READ:
                deniedPermissions.add(WRITE_SMS);
                break;
        }

        doCheck(deniedPermissions);
    }

    @SuppressWarnings({"deprecation", "RedundantSuppression"}) // TODO: 06.02.2020 deprecated
    protected void onPermissionsDenied(Collection<String> permissions) {
        log.debug("Denied: " + permissions);

        if (!permissions.isEmpty()) {
            SharedPreferencesWrapper.EditorWrapper edit = settings.edit();

            for (String permission : permissions) {
                switch (permission) {
                    case RECEIVE_SMS:
                        edit.removeFromStringSet(PREF_EMAIL_TRIGGERS, VAL_PREF_TRIGGER_IN_SMS);
                        break;
                    case READ_SMS:
                        edit.removeFromStringSet(PREF_EMAIL_TRIGGERS, VAL_PREF_TRIGGER_OUT_SMS);
                        break;
                    case WRITE_SMS:
                        edit.removeFromStringSet(PREF_MARK_SMS_AS_READ);
                        break;
                    case READ_PHONE_STATE:
                        edit.removeFromStringSet(PREF_EMAIL_TRIGGERS, VAL_PREF_TRIGGER_IN_CALLS, VAL_PREF_TRIGGER_MISSED_CALLS);
                        break;
                    case PROCESS_OUTGOING_CALLS:
                        edit.removeFromStringSet(PREF_EMAIL_TRIGGERS, VAL_PREF_TRIGGER_OUT_CALLS);
                        break;
                    case READ_CONTACTS:
                        edit.removeFromStringSet(PREF_EMAIL_CONTENT, VAL_PREF_EMAIL_CONTENT_CONTACT);
                        break;
                    case ACCESS_COARSE_LOCATION:
                    case ACCESS_FINE_LOCATION:
                        edit.removeFromStringSet(PREF_EMAIL_CONTENT, VAL_PREF_EMAIL_CONTENT_LOCATION);
                        break;
                }
            }

            edit.apply();
        }
    }

    private void doCheck(Collection<String> permissions) {
        if (!permissions.isEmpty()) {
            List<String> deniedPermissions = new ArrayList<>();
            List<String> unexplainedPermissions = new ArrayList<>();

            for (String permission : permissions) {
                if (checkSelfPermission(activity, permission) != PERMISSION_GRANTED) {
                    deniedPermissions.add(permission);
                    if (shouldShowRequestPermissionRationale(activity, permission)) {
                        unexplainedPermissions.add(permission);
                    }
                }
            }

            if (!deniedPermissions.isEmpty()) {
                if (!unexplainedPermissions.isEmpty()) {
                    explainPermissions(unexplainedPermissions);
                } else {
                    requestPermissions(deniedPermissions);
                }
            }
        }
    }

    private void requestPermissions(List<String> permissions) {
        log.debug("Requesting : " + permissions);

        ActivityCompat.requestPermissions(activity, toArray(permissions, String.class), requestResultCode);
    }

    private void explainPermissions(final List<String> permissions) {
        log.debug("Explaining : " + permissions);

        new AlertDialog.Builder(activity)
                .setMessage(formatRationale(permissions))
                .setPositiveButton(android.R.string.ok, (dialog, which) -> requestPermissions(permissions))
                .show();
    }

    private String formatRationale(Collection<String> permissions) {
        StringBuilder b = new StringBuilder();
        for (String permission : permissions) {
            String line = formatter(activity)
                    .pattern(Objects.requireNonNull(items.get(permission)))
                    .put("permission", getPermissionLabel(permission))
                    .format();
            b.append(line).append("\n\n");
        }
        return b.toString();
    }

    private String getPermissionLabel(String permission) {
        try {
            PackageManager packageManager = activity.getPackageManager();
            PermissionInfo info = packageManager.getPermissionInfo(permission, 0);
            return info.loadLabel(packageManager).toString();
        } catch (PackageManager.NameNotFoundException x) {
            throw new RuntimeException(x);
        }
    }

    private void showDenialImpact() {
        new AlertDialog.Builder(activity)
                .setMessage(R.string.since_permissions_not_granted)
                .setPositiveButton(android.R.string.ok, null)
                .show();
    }

    static boolean isLocationPermissionsGranted(Context context) {
        return checkPermission(context, ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION);
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/PermissionsHelper.kt
package com.bopr.android.smailer

import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.Manifest.permission.POST_NOTIFICATIONS
import android.Manifest.permission.READ_CALL_LOG
import android.Manifest.permission.READ_CONTACTS
import android.Manifest.permission.READ_PHONE_STATE
import android.Manifest.permission.READ_SMS
import android.Manifest.permission.RECEIVE_SMS
import android.Manifest.permission.SEND_SMS
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Build
import androidx.activity.result.contract.ActivityResultContracts.RequestMultiplePermissions
import androidx.core.app.ActivityCompat.shouldShowRequestPermissionRationale
import androidx.core.content.ContextCompat.checkSelfPermission
import androidx.fragment.app.FragmentActivity
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_MESSAGE_CONTENT
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_SENDER_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_TRIGGERS
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_MESSENGER_RECIPIENTS
import com.bopr.android.smailer.Settings.Companion.PREF_REMOTE_CONTROL_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_MESSAGE_CONTENT_CALLER
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_MESSAGE_CONTENT_LOCATION
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_TRIGGER_IN_CALLS
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_TRIGGER_IN_SMS
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_TRIGGER_MISSED_CALLS
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_TRIGGER_OUT_CALLS
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_TRIGGER_OUT_SMS
import com.bopr.android.smailer.ui.InfoDialog.Companion.showInfoDialog
import org.slf4j.LoggerFactory

/**
 * Responsible for permissions checking.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
class PermissionsHelper(
    private val activity: FragmentActivity,
    private val onPermissionsRequestComplete: () -> Unit = {}
) {

    private val items: Map<String, Int> = sortedMapOf<String, Int>().apply {
        put(RECEIVE_SMS, R.string.permission_rationale_receive_sms)
        put(RECEIVE_SMS, R.string.permission_rationale_receive_sms)
        put(SEND_SMS, R.string.permission_rationale_send_sms)
        put(READ_SMS, R.string.permission_rationale_read_sms)
        put(READ_PHONE_STATE, R.string.permission_rationale_phone_state)
        put(READ_CALL_LOG, R.string.permission_rationale_phone_state)
        put(READ_CONTACTS, R.string.permission_rationale_read_contacts)
        put(ACCESS_COARSE_LOCATION, R.string.permission_rationale_coarse_location)
        put(ACCESS_FINE_LOCATION, R.string.permission_rationale_fine_location)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            put(POST_NOTIFICATIONS, R.string.permission_rationale_post_notifications)
        }
    }

    private val settings = Settings(activity)
    private val accountHelper = AccountHelper(activity)
    private val permissionRequestLauncher =
        activity.registerForActivityResult(RequestMultiplePermissions()) { result ->
            onPermissionRequestResult(result)
        }

    fun checkAll() {
        checkPermissions(items.keys)
    }

    private fun onPermissionRequestResult(result: Map<String, @JvmSuppressWildcards Boolean>) {
        val grantedPermissions = result.filterValues { it }.keys
        val deniedPermissions = result.filterValues { !it }.keys

        onPermissionsGranted(grantedPermissions)
        onPermissionsDenied(deniedPermissions)

        if (deniedPermissions.isNotEmpty()) {
            activity.showInfoDialog(messageRes = R.string.since_permissions_not_granted)
        } else {
            onPermissionsRequestComplete()
        }
    }

    /**
     * To be added into owner's onSharedPreferenceChanged()
     */
    internal fun onSettingsChanged(key: String?) {
        log.debug("Handling preference changed: $key")

        val requiredPermissions = mutableSetOf<String>()
        when (key) {
            PREF_EMAIL_TRIGGERS -> {
                val triggers = settings.getEmailTriggers()
                if (triggers.contains(VAL_PREF_TRIGGER_IN_SMS)) {
                    requiredPermissions.add(RECEIVE_SMS)
                }
                if (triggers.contains(VAL_PREF_TRIGGER_OUT_SMS)) {
                    requiredPermissions.add(READ_SMS)
                }
                if (triggers.contains(VAL_PREF_TRIGGER_IN_CALLS)
                    || triggers.contains(VAL_PREF_TRIGGER_OUT_CALLS)
                    || triggers.contains(VAL_PREF_TRIGGER_MISSED_CALLS)
                ) {
                    requiredPermissions.add(READ_PHONE_STATE)
                    requiredPermissions.add(READ_CALL_LOG)
                }
            }

            PREF_EMAIL_MESSAGE_CONTENT -> {
                val content = settings.getStringSet(PREF_EMAIL_MESSAGE_CONTENT)
                if (content.contains(VAL_PREF_MESSAGE_CONTENT_CALLER)) {
                    requiredPermissions.add(READ_CONTACTS)
                }
                if (content.contains(VAL_PREF_MESSAGE_CONTENT_LOCATION)) {
                    requiredPermissions.add(ACCESS_COARSE_LOCATION)
                    requiredPermissions.add(ACCESS_FINE_LOCATION)
                }
            }
        }

        checkPermissions(requiredPermissions)
    }

    private fun onPermissionsGranted(grantedPermissions: Set<String>) {
        log.debug("Granted: {}", grantedPermissions)

        /* set default accounts at startup */
        val accountName = accountHelper.getPrimaryGoogleAccount()?.name
        settings.update {
            ifNotExists(PREF_EMAIL_SENDER_ACCOUNT) { putString(it, accountName) }
            ifNotExists(PREF_EMAIL_MESSENGER_RECIPIENTS) { putString(it, accountName) }
            ifNotExists(PREF_REMOTE_CONTROL_ACCOUNT) { putString(it, accountName) }
        }
    }

    private fun onPermissionsDenied(deniedPermissions: Collection<String>) {
        log.debug("Denied: {}", deniedPermissions)

        if (deniedPermissions.isNotEmpty()) {
            val triggers = settings.getEmailTriggers()
            val content = settings.getStringSet(PREF_EMAIL_MESSAGE_CONTENT)

            if (deniedPermissions.contains(RECEIVE_SMS)) {
                triggers.remove(VAL_PREF_TRIGGER_IN_SMS)
            }
            if (deniedPermissions.contains(READ_SMS)) {
                triggers.remove(VAL_PREF_TRIGGER_OUT_SMS)
            }
            if (deniedPermissions.contains(READ_PHONE_STATE)
                || deniedPermissions.contains(READ_CALL_LOG)
            ) {
                triggers.remove(VAL_PREF_TRIGGER_IN_CALLS)
                triggers.remove(VAL_PREF_TRIGGER_OUT_CALLS)
                triggers.remove(VAL_PREF_TRIGGER_MISSED_CALLS)
            }
            if (deniedPermissions.contains(READ_CONTACTS)) {
                content.remove(VAL_PREF_MESSAGE_CONTENT_CALLER)
            }
            if (deniedPermissions.contains(ACCESS_COARSE_LOCATION)
                || deniedPermissions.contains(ACCESS_FINE_LOCATION)
            ) {
                content.remove(VAL_PREF_MESSAGE_CONTENT_LOCATION)
            }

            settings.update {
                putStringSet(PREF_EMAIL_TRIGGERS, triggers)
                putStringSet(PREF_EMAIL_MESSAGE_CONTENT, content)
            }
        }
    }

    private fun checkPermissions(permissions: Collection<String>) {
        if (permissions.isEmpty()) {
            onPermissionsRequestComplete()
        } else {
            log.debug("Checking: {}", permissions)

            val deniedPermissions = mutableListOf<String>()
            val unexplainedPermissions = mutableListOf<String>()

            for (p in permissions) {
                if (checkSelfPermission(activity, p) != PERMISSION_GRANTED) {
                    deniedPermissions.add(p)
                    if (shouldShowRequestPermissionRationale(activity, p)) {
                        unexplainedPermissions.add(p)
                    }
                }
            }

            if (deniedPermissions.isEmpty()) {
                onPermissionsRequestComplete()
            } else {
                if (unexplainedPermissions.isNotEmpty()) {
                    explainPermissions(unexplainedPermissions)
                } else {
                    requestPermissions(deniedPermissions)
                }
            }
        }
    }

    private fun requestPermissions(permissions: List<String>) {
        log.debug("Requesting : {}", permissions)
        permissionRequestLauncher.launch(permissions.toTypedArray())
    }

    private fun explainPermissions(permissions: List<String>) {
        log.debug("Explaining : {}", permissions)

        activity.showInfoDialog(message = formatRationale(permissions)) {
            requestPermissions(permissions)
        }
    }

    private fun formatRationale(permissions: Collection<String>): String {
        return buildString {
            for (permission in permissions) {
                val value = items.getValue(permission)
                append(activity.getString(value, getLabelFor(permission)))
                append("\n\n")
            }
        }
    }

    private fun getLabelFor(permissionName: String): String {
        return activity.packageManager.run {
            getPermissionInfo(permissionName, 0).loadLabel(this).toString()
        }
    }

    companion object {


        private val log = LoggerFactory.getLogger("PermissionsHelper")
    }

}